Add Semaphore 2.0 integration.

abhinav824 opened the pull request:
Semaphore has two versions operating as two distinct products and currently zulip supports only the older version.
This commit extends support for Semaphore 2.0.

Tested on my local development server. Integration yet to be tested on the Semaphore 2.0 end.

Fixes #14171

punchagan commented:
@abhinav824 Thanks for working on this. Could you fix the linting issues that cause CI to fail, please?

punchagan commented on zerver/webhooks/semaphore2/view.py:
```diff
@@ -0,0 +1,166 @@
+* **Pull Request**: [{pull_request_title}]({pull_request_url})
+* **Pipeline**:
+  - Name: {pipeline_name}
+  - State: {pipeline_state}
```
At what point is the Pull request event triggered? Is it always when the PR is opened, in which case the state would always be the same, and probably not very useful. Similarly, for the push event too.

abhinav824 replied:
I found [this](https://docs.semaphoreci.com/essentials/project-workflow-trigger-options/) page, which shows that every commit to a pull request triggers a pipeline/workflow and notifications are triggered at the completion (success or failure) of a pipeline.
I think you're right, in all the cases state will always be 'stopped'. Should i remove the state ?

punchagan replied:
Yes, I think it can be removed.

abhinav824 replied:
@punchagan I think we can use `result_reason` field for pipeline and blocks. This field will be useful to track down any issue, if the pipeline stops due to some error. There are [multiple](https://docs.semaphoreci.com/reference/pipeline-yaml-reference/#result_reason) possible values for this field which tells the reason of particular `result` value.

abhinav824 commented on zerver/webhooks/semaphore2/view.py:
```diff
@@ -0,0 +1,166 @@
+        Returns the url of a particular commit using the commit_id and
+        repository url.
+    """
+    return repo_url + '/commit/' + commit_id
```
Thanks for pointing this out. I have gone through the documentation of semaphore. I think, as of now they only support Github and Github enterprise (have a look on [this](https://docs.semaphoreci.com/account-management/working-with-github-sso/)). I'm also not sure whether Enterprise Github URLs follow the same pattern as that of Github.

What we can do is, we can have a dictionary of supported repository hosting services and their URL pattern and use it for building commit/PR URL. For the services which are not in that dictionary (we cannot build URLs for them), we can redirect the user to the repository itself or we can just print out the commit hash (or the PR id).

punchagan replied:
I would expect Semaphore to add these URLs to their payload, once they start supporting other services. But, for now, what you suggest seems reasonable.

abhinav824 commented:
@punchagan I've fixed the linting issues but, the CI is still failing because of following cases which are not covered in tests:
1. We've considered only two events as of now i.e push or a pull request. In the `if` condition, i've also kept an `else` block which will work if the event is not a 'push' or a 'pull_request' (which can never be a case). should i remove that block or should i add a test for it or should i add a non-coverage tag ? 
1. Helper function which returns the commit/PR URLs also contains an `else` block which will be executed if the repository host URL is not in the list of known hosts (i.e Github). Should i write a test for this case or should i add a non-coverage comment for this ?

punchagan commented on zerver/webhooks/semaphore2/view.py:
```diff
@@ -0,0 +1,184 @@
+    # New hosting services must be added when semaphore extends it support.
+
+    repo_slug = '/' + repo_slug
+    repository_hosting_service_url = repo_url.replace(repo_slug, '')
```
I would parse the URL and get the domain name/host, instead of this string manipulation.

abhinav824 replied:
I'll use `urllib.parse` and get the domain name, that would be a lot better. Thanks!

punchagan commented on zerver/webhooks/semaphore2/view.py:
```diff
@@ -0,0 +1,184 @@
+    BLOCKS_DATA = """"""
+    BLOCK_BODY = """
+**1. {name}**:
+  - State: {state}
```
I think I'd remove the state for the Blocks too.

abhinav824 replied:
I thought the state of blocks might be useful as a pipeline consists of multiple blocks. But, result would suffice and state may be same for all the cases. So, i'll remove the state for blocks too.

punchagan commented:
> * We've considered only two events as of now i.e push or a pull request. In the `if` condition, i've also kept an `else` block which will work if the event is not a 'push' or a 'pull_request' (which can never be a case). should i remove that block or should i add a test for it or should i add a non-coverage tag ?

If we are never getting to that branch (based on how you are setting the event variable), it wouldn't be easy to write a test for this. I think we can ignore coverage for this. 

> * Helper function which returns the commit/PR URLs also contains an `else` block which will be executed if the repository host URL is not in the list of known hosts (i.e Github). Should i write a test for this case or should i add a non-coverage comment for this ?

I think we should have a test for this.

abhinav824 commented:
@punchagan I've made the necessary changes, please have a look.

punchagan commented on zerver/webhooks/semaphore2/tests.py:
```diff
@@ -0,0 +1,74 @@
+  - Result: stopped
+""".strip()
+        self.send_and_test_stream_message('push_unknown_host', expected_topic, expected_message,
+                                          content_type="application/x-www-form-urlencoded")
```
Is the `content_type` really `application/x-www-form-urlencoded`? I would expect it to be `application/json`.

abhinav824 replied:
I also think it should be `application/json`. I just used the existing semaphore integration for adding this integration. I'll change it to `application/json` for semaphore2.

punchagan commented on zerver/webhooks/semaphore2/view.py:
```diff
@@ -0,0 +1,180 @@
+            blocks = blocks,
+        )
+
+    else:  # nocoverage
```
It would still be useful to retain the comment here that we never get to this branch.

abhinav824 replied:
Okay, I'll add that comment.

punchagan commented on zerver/webhooks/semaphore2/view.py:
```diff
@@ -0,0 +1,180 @@
+    branch_name = payload["revision"]["branch"]["name"]
+    blocks = get_blocks_data(payload["blocks"])
+
+    if payload["revision"]["pull_request"] is None:
```
I looked at the semaphore docs again, and they say

> In this example `revision.pull_request` and `revision.tag` are null because payload is related to the pipeline run started from a push to the branch. Information about this is kept in `revision.reference_type`.

The documentation is not clear on what the possible values for `reference_type` are, but I suspect it would be different for a `pull_request`. You might have to play with Semaphore to figure this out. 

Also, it sounds like it is possible to have events triggered when a `tag` is pushed, because there's a `revision.tag` field they mention? So, this check on the payload may not be sufficient, and we can do better, I think.

abhinav824 replied:
I think you're right. I just ignored the `revision.tag` field, there must be some significance of it. I'll try to look deeper into the semaphore docs and try to figure out every possible event which can trigger a notification.

punchagan commented on zerver/webhooks/semaphore2/view.py:
```diff
@@ -0,0 +1,180 @@
+        name = block["name"].title()
+        result = block["result"]
+        content = BLOCK_BODY.format(name=name, result=result)
+        BLOCKS_DATA += content
```
Would the markdown be concatenated correctly, if there are multiple blocks? It might be worth adding multiple blocks to one of the tests.

abhinav824 replied:
Thanks for pointing this out. I'll add multiple blocks to one of a tests and fixtures.

punchagan commented on zerver/webhooks/semaphore2/view.py:
```diff
@@ -0,0 +1,180 @@
+# a pipeline is triggered.
+
+PUSH_TEMPLATE = """
+**{email}** pushed to branch **{branch_name}** of \
```
Would it be better to use the login/username instead of the email? The payload doesn't seem to have name.

abhinav824 replied:
I wasn't sure that login is username or the name of the user. We just have to use an unique field and if login is the username of the user then we can definitely use that.

punchagan replied:
I suspect it would be the username. You could play around with the Semaphore API and verify.

punchagan commented on zerver/webhooks/semaphore2/view.py:
```diff
@@ -0,0 +1,180 @@
+
+    return BLOCKS_DATA
+
+# Funtions to generate url to a commit/Pull Request. This builds an url
```
This comment is not necessary, I think. The function names and doc strings below are good enough.

abhinav824 replied:
Okay, I'll remove that comment.

punchagan commented on zerver/webhooks/semaphore2/view.py:
```diff
@@ -0,0 +1,180 @@
+from zerver.models import UserProfile
+from urllib.parse import urlparse
+
+# There seems to be a possibility of only 2 types of events : Push or a
```
This comment would need to be changed, if indeed there's a `tag` related event. See comment below.

punchagan commented:
> @punchagan I've made the necessary changes, please have a look.

Thanks again, for working on this, @abhinav824! 

I reviewed the PR more carefully this time, also while referencing the Semaphore docs. I'm not sure any more that Semaphore has only two kinds of events. It would be great if you could investigate this.

abhinav824 commented on zerver/lib/integrations.py:
```diff
@@ -327,6 +327,7 @@ def __init__(self, name: str, *args: Any, **kwargs: Any) -> None:
     WebhookIntegration('pivotal', ['project-management'], display_name='Pivotal Tracker'),
     WebhookIntegration('raygun', ['monitoring'], display_name="Raygun"),
     WebhookIntegration('reviewboard', ['version-control'], display_name="ReviewBoard"),
+    WebhookIntegration('semaphore2', ['continuous-integration', 'deployment'], stream_name='pipelines'),
```
@punchagan Should i add docstring to `WebhookIntegration` ? (as per the 1st point)

abhinav824 replied:
@punchagan could you please elaborate the first point. I mean i don't understand where exactly i need to add docs ?

punchagan replied:
I'm also not entirely sure, but looking at the `stream_name` argument, I couldn't immediately figure out what it actually does. May be on the `Integration` class or the `WebhookIntegration` class? It can come as a separate commit.

abhinav824 commented:
@punchagan I've made the following changes after looking into the documentaion of semaphore and Github.

1.  Replaced the email with login (i.e username) in the notification. I've looked into the github webhook [documentation](https://developer.github.com/v3/activity/events/types/#pushevent). I believe that `login` refers to the username of the user (If you look into any payload in github documentation and semaphore payload, both of them have `sender` field which contains `login` and `email` field). I'm also building the link of user profile (same as we did for commit/PR urls).
1.  Added a `tag` event and used `revision.reference_type` field for identifying the event. The possible values of `reference_type` are given in this [page](https://docs.semaphoreci.com/essentials/project-workflow-trigger-options/). This page also shows that there can be only four type of events which can trigger a workflow/pipeline.

Apart from these, some small changes are also there.
1. Changed comments accordingly.
1. Added multiple blocks in the `push.json` fixture to make sure that multiple blocks are getting appended correctly into the notification content.
1. Changed `content_type` from `application/x-www-form-urlencoded` to `application/json` in `tests.py`.
1. Removed `stream_name` for both of the Semaphore integrations in `WebhookIntegration`.

I've tried to fix all the issues that were mentioned above, Please have a look. 

For the `stream_name` issue, currently i'm not sure how does it work. So, I need to some time to look into it. I'll push any changes related to it in a separate commit.

timabbott commented:
Looking at this, one question I have is whether we need a separate `semaphore2` integration, or if we can detect which format of Semaphore request we're getting by looking at headers or keys, and avoid changing the URL?  Generally we prefer doing so whenever possible, and it's almost certainly possible here.

punchagan commented:
We (@abhinav824 and I) did discuss this in the original issue, and weren't sure which way to go. As an end-user it does make life easier for me if there was just one integration which handled both versions. We ended up going the separate route because it would make the documentation simpler, and allow for displaying the two slightly different logos, etc. 

The Semaphore site itself doesn't mention that the current version is Semaphore 2.0, and it might be confusing for new Semaphore users which integration they should pick. If we did go with two integrations, we should rename the older one to Semaphore Classic, probably.  

But, if we have a general preference towards having a single integration, we should probably merge the two, and I don't think it would be super tricky to do so. @abhinav824 would you be happy to go ahead and merge the two integrations? Or do you have other thoughts on this?

abhinav824 commented:
@punchagan @timabbott  I think we should keep two separate integrations. The two versions of semaphore differs a lot and it would be better if we keep them separate at our end also. The notification content we're sending in both the cases, the pipeline concept, payload, etc makes it significantly different. Even they've also mentioned on their website that the two versions of semaphore are [two distinct products](https://docs.semaphoreci.com/guided-tour/migration-guide-for-semaphore-classic-users/#semaphore-classic-and-semaphore-20-are-distinct-products). What i suggest is, we should keep them separate and rename to older version to 'Semaphore Classic' and add a note in both (Semahore Classic and Semaphore 2.0) the documentations to avoid confusion at user end. I don't think user will find much difficulty in identifying the version of semaphore as, the logos for both the versions are different.

timabbott commented:
My thinking is mainly that they're certain to shut down the old integration, and it'd be nice to not end up with 2 URLs.  And we have users who have already pointed Semaphore 2 at our `/semaphore` webhook URL, and I'd really like to just have things start working rather than needing to get them to change the URL.

Which is another example of the one of reasoning that minimizing complexity for the user is often the right thing to prioritize.  So I think we should migrate.

abhinav824 commented:
Since the users already pointed Semaphore 2 at `/semaphore` webhook, then i think it would be better to minimise the user effort. I'll merge the two integrations.

timabbott commented:
@abhinav824 let us know when you have a merged version ready.  It'd be really nice to integrate this in the next few days.

punchagan commented:
@abhinav824 do let me know if you need any help with this, I'd be more than happy to help!

abhinav824 commented:
Sure, I'll let you know when I'm done with it. I've already merged the two versions but somehow, my development environment broke and I'm not able to test it. Also, I've some upcoming projects (college) submission this weak. So, I'll try to fix the issue and push the changes as soon as possible.

timabbott commented:
@abhinav824 can you push what you have?  I'd really like to eliminate the 500s from this on zulipchat.com, since they clutter error logs, so it might make sense for someone else to take over finishing this from here.

punchagan commented:
(I've turned off my Semaphore webhook, though it probably doesn't help that much with reducing the clutter. :))

abhinav824 commented:
@punchagan @timabbott I've merged the two integrations. Please have a look.
There is an already existing fixture `eption_message.json` (for Semaphore Classic) which produces internal server error (many fields are missing) when we try to trigger notification using this fixture (using devtools). Since there is a comment that it'll never be executed, I haven't fixed it. If you want I can fix it but I have some small doubts regarding it that I'll mention in the next comment.

abhinav824 commented on zerver/webhooks/semaphore/view.py:
```diff
@@ -22,59 +25,186 @@
+    if 'event' in payload.keys():
+        # semaphore only gives the last commit, even if there were multiple commits
+        # since the last build
+        branch_name = payload["branch_name"]
+        project_name = payload["project_name"]
+        result = payload["result"]
+        event = payload["event"]
+        commit_id = payload["commit"]["id"]
+        commit_url = payload["commit"]["url"]
+        author_email = payload["commit"]["author_email"]
+        message = payload["commit"]["message"]
```
There are many missing fields which are producing `missing_key` errors, should I assign default values in case of missing field or should I change the notification content so that these variables are not used and shift the statements fetching these variables from payload inside the existing `if` conditions which are using these variables.

abhinav824 commented on zerver/webhooks/semaphore/view.py:
```diff
@@ -22,59 +25,186 @@
+                server_name=server_name
+            )
+
+        else:  # should never get here
+            content = "{event}: {result}".format(
+                event=event, result=result)
```
This `else` condition should be executed when we use `eption_message.json` fixture.

punchagan commented:
Thanks @abhinav824 for pushing your changes. @timabbott I'm happy to take this up, and wrap it up. I need this for a Zulip Realm that I'm part of, and would be happy to wrap it up sooner, rather than later.

timabbott commented:
Closing now that #14524 is merged.  Thanks for all your work on this @abhinav824!

abhinav824 commented:
Thanks @punchagan @timabbott for helping me out in my very first contribution to zulip.
